### DISCUSSION TITLE
ActivityPub: Automatically enable bridge for all users on a given instance

### DISCUSSION THREAD

**Discussion by TomCasavant** [snarfed/bridgy-fed#1118] [General]
I swear I saw a github issue about making bluesky opt-out where I would've asked this but I can't find it anymore

But, assuming you do go the route of making activitypub opt-in and bluesky opt-out is there any consideration for making a specific server (specifically threads) opt-out as well? Just because the culture there seems more open to the idea of federating with bluesky than the rest of the social-web is (though they'd still have to opt-in to sharing with the fediverse anyways)

**Comment by snarfed** [comment]
Hoo boy! That's a big one. Honestly I have no idea. Other instance admins have already asked for this, so it's definitely possible.

One catch is that right now, at least the way fediverse software is generally designed, I'd still need to have the `@bsky.brid.gy@bsky.brid.gy` bot user follow each user on an auto-enabled instance like this, so that it delivers their posts to BF. So, they need some way to automatically have every new user follow the bot user, so that it will follow them back. It sounds like this is maybe possible in at least some software, eg Mastodon...

...but if that's necessary, and they do that, then there's nothing special I'd need to do on BF's end.

(The alternative would be for fediverse software to start adding an option to automatically deliver all posts to a given remote inbox. Maybe some can already do this? I don't know, but seems unlikely.)

**Reply by TomCasavant** [reply]
I was thinking that it'd only bridge the user if someone on bluesky searched for and followed them (or whenever the bridge discovers the user via a boost by a bridged user) rather than immediately bridging every user (which I think is how following websites works?). That'd still allow the bot to still individually follow each user

**Reply by snarfed** [reply]
I wish! We can do this for bridging into the fediverse because searching on a (logged in) fediverse instance makes outbound webfinger requests to remote servers. Sadly Bluesky search runs entirely on the appview's index, it doesn't make any remote requests that we can hook into. More in https://github.com/bluesky-social/atproto/discussions/1650 . So to get a bridged account to show up in Bluesky search, we have to proactively bridge it and push it there ahead of time.

**Reply by qazmlp** [reply]
> (The alternative would be for fediverse software to start adding an option to automatically deliver all posts to a given remote inbox. Maybe some can already do this? I don't know, but seems unlikely.)

That's an ActivityPub Relay connection, essentially. Gets all Public posts and is supported cross-app to some extent (at least by Mastodon and Pixelfed, but I assume there are more).

I don't know the specifics of the API, but if Bridgy Fed had a Relay endpoint, that could have this exact behaviour if the instance's admins choose to connect to it. You would also still receive relevant mentions and likes regardless.

(I'm not 100% sure on this because I couldn't find the API docs, but *I think* a Relay connection comes with enough authentication for this use-case.)

The downside is that blocking the respective Bridgy Fed domain wouldn't act as opt-out. I'm also not sure you can share the same inbox between multiple Relay domains to deduplicate incoming activities.

**Reply by qazmlp** [reply]
I did a bit of research and it's fairly simple (a bit *too* simple, really):

Relays are specialised bot accounts (usually `@relay@domain`) that present at least an inbox/shared inbox.
To send public activity to a relay, Pleroma admins add that actor in their settings (by ID, it seems), while Mastodon admins add the inbox directly. I think joining a relay may also involve following that actor somehow, but I assume it's not just any user who can do that.

There's a bit more to it in terms of sending the activities back, but that's mostly not relevant to Bridgy Fed since the primary use of Relays is to join topically-related instances. (Bridgy is likely a bit too general, so it's unlikely that a fedi instance would want *all* posts.)

There are relays out there that offer hashtag subscriptions though, which may be interesting for Bridgy Fed. For example, https://relay-new.activitypub.ca explains in its docs that instance admins can DM that relay something like

> ```
> @relay@domain subscribe-hashtag #hashtag1 #hashtag2
> ```

to get all posts that mention any of these hashtags. (Normally relays have to `Announce` new posts since they aren't authoritative, though depending on the subscribers Bridgy *may* be able to just send the activities as normal.)

I'm going to dig a bit more into what actions are implied when joining a relay from Mastodon/Pleroma.
(Edit: It requires following the relay actor and for that follow to be accepted to enable the outbound connection too, I think. I'm not sure which actor is used to follow the relay. I don't know how Mastodon determines the relay actor from the inbox URI. Maybe it just assumes `@relay@domain`?)

**Reply by qazmlp** [reply]
Another aspect that would be interesting for use as a relay would be subscribing to certain custom feeds.
Those would be indistinct on the "Federated timeline", but with many feeds it would be like joining a certain topic.
At that point you may also want to consider `Announce`ing posts that appear in the custom feed but were originally sent from AP through Bridgy Fed. (For a good experience, that would also require `@id`-only-forwarding for activities that can't be `Announce`d normally, like I think `Update`s.)

That's definitely something I'd consider a far-future feature, though.

**Reply by snarfed** [reply]
Thanks for all the info here!

By "custom feeds," do you mean Bluesky custom feeds? More on that in #1095

**Reply by qazmlp** [reply]
> By "custom feeds," do you mean Bluesky custom feeds? More on that in #1095

Yes, and the same challenges apply. (I don't think all custom feeds can be translated into AP `Announce` streams, since some at least change every time you view them.)

**Comment by snarfed** [comment]
Corresponding issue is #1305
